### PR TITLE
Ruby: flow through instance variables

### PR DESCRIPTION
**Pull request by aibaars**

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/ast/Variable.qll [hunk 1]
```diff
@@ -181,6 +181,17 @@ class GlobalVariableReadAccess extends GlobalVariableAccess, VariableReadAccess
 /** An access to an instance variable. */
 class InstanceVariableAccess extends VariableAccess instanceof InstanceVariableAccessImpl {
   final override string getAPrimaryQlClass() { result = "InstanceVariableAccess" }
+
+  /**
+   * Gets a synthetic `self` variable access.
+   */
+  final SelfVariableAccess getSelfVariableAccess() { synthChild(this, 0, result) }
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/ast/Variable.qll:188]: Should we call it `getReceiver`, like in method calls?

File: ruby/ql/lib/codeql/ruby/controlflow/CfgNodes.qll [hunk 2]
```diff
@@ -615,6 +615,22 @@ module ExprNodes {
     final override VariableReadAccess getExpr() { result = ExprCfgNode.super.getExpr() }
   }
 
+  private class InstanceVariableAccessMapping extends ExprChildMapping, InstanceVariableAccess {
+    override predicate relevantChild(AstNode n) { n = this.getSelfVariableAccess() }
+  }
+
+  /** A control-flow node that wraps a `InstanceVariableAccess` AST expression. */
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/controlflow/CfgNodes.qll:622]: an

File: ruby/ql/lib/codeql/ruby/controlflow/CfgNodes.qll [hunk 3]
```diff
@@ -615,6 +615,22 @@ module ExprNodes {
...
+  class InstanceVariableAccessCfgNode extends ExprCfgNode {
+    override InstanceVariableAccessMapping e;
+
+    final override InstanceVariableAccess getExpr() { result = ExprCfgNode.super.getExpr() }
+
+    /**
+     * Gets a synthetic `self` variable access.
+     */
+    final CfgNode getSelfVariableAccess() { e.hasCfgChild(e.getSelfVariableAccess(), this, result) }
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/controlflow/CfgNodes.qll:631]: Also rename this, if renaming the one on the AST class.

File: ruby/ql/test/library-tests/dataflow/global/instance_variables.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+class Foo
+    def set_field x
+        @field = x
+    end
+    def get_field
+        return @field
+    end
+    def inc_field
```
Comment (hvitved) [inline ruby/ql/test/library-tests/dataflow/global/instance_variables.rb:8]: not used?

**Review by hvitved — COMMENTED** [review]
LGTM, some minor comments. I'll submit a small performance fix. This PR does not add support for fields in flow summaries/models-as-data; is this something we should add (perhaps as follow-up)?

File: ruby/ql/test/library-tests/dataflow/global/instance_variables.rb (hunk 4 above)
Comment (aibaars) [reply ruby/ql/test/library-tests/dataflow/global/instance_variables.rb:8]: I added a taint flow test now.

File: ruby/ql/lib/codeql/ruby/controlflow/CfgNodes.qll (hunk 2 above)
Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/controlflow/CfgNodes.qll]: Done, I also spotted another case like this for `InstanceVariableWriteAccessCfgNode` which made me realise I should probably use that class in the `storeStep` implementation.

File: ruby/ql/lib/codeql/ruby/ast/Variable.qll [hunk 5]
```diff
@@ -181,6 +181,17 @@ class GlobalVariableReadAccess extends GlobalVariableAccess, VariableReadAccess
 /** An access to an instance variable. */
 class InstanceVariableAccess extends VariableAccess instanceof InstanceVariableAccessImpl {
   final override string getAPrimaryQlClass() { result = "InstanceVariableAccess" }
+
+  /**
+   * Gets the synthetic receiver(`self`) of this instance variable access.
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/ast/Variable.qll:186]:
```suggestion
   * Gets the synthetic receiver (`self`) of this instance variable access.
```

File: ruby/ql/lib/codeql/ruby/controlflow/CfgNodes.qll [hunk 6]
```diff
@@ -615,6 +615,22 @@ module ExprNodes {
...
+  }
+
+  /** A control-flow node that wraps an `InstanceVariableAccess` AST expression. */
+  class InstanceVariableAccessCfgNode extends ExprCfgNode {
+    override InstanceVariableAccessMapping e;
+
+    override InstanceVariableAccess getExpr() { result = ExprCfgNode.super.getExpr() }
+
+    /**
+     * Gets the synthetic receiver(`self`) of this instance variable access.
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/controlflow/CfgNodes.qll:629]:
```suggestion
     * Gets the synthetic receiver (`self`) of this instance variable access.
```

**Review by hvitved — APPROVED** [review]
Performance looks good now.

**Pull request merged**